pkg/dev/dev.go

```diff
@@ -74,3 +96,107 @@ func (o *DevClient) Cleanup() error {
 	var err error
 	return err
 }
+
+// SetupPortForwarding sets up port forwarding for the endpoints in the devfile
+func (o *DevClient) SetupPortForwarding(devfileObj parser.DevfileObj, envSpecificInfo *envinfo.EnvSpecificInfo, out, errOut io.Writer) error {
+	var err error
+	var containers []v1.Component
+	containers, err = devfileObj.Data.GetComponents(parsercommon.DevfileOptions{
+		ComponentOptions: parsercommon.ComponentOptions{ComponentType: v1.ContainerComponentType},
+	})
+
+	endpoints := endpointsFromContainers(containers)
+	portPairs := portPairsFromEndpoints(endpoints)
+
+	if len(portPairs) == 0 {
+		// no endpoints with exposure set to public or internal; no ports to be forwarded
+		return nil
+	}
+
+	var pod *corev1.Pod
+	pod, err = o.kubernetesClient.GetOnePodFromSelector(componentlabels.GetSelector(devfileObj.GetMetadataName(), envSpecificInfo.GetApplication()))
+	if err != nil {
+		return err
+	}
+
+	transport, upgrader, err := spdy.RoundTripperFor(o.kubernetesClient.GetClientConfig())
+	if err != nil {
+		return err
+	}
+
+	req := o.kubernetesClient.GeneratePortForwardReq(pod.Name)
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", req.URL())
+	stopChan := make(chan struct{}, 1)
+	// passing nil for readyChan because it's eventually being closed if it's not nil
+	// passing nil for out because we only care for error, not for output messages; we want to print our own messages
+	var fw *portforward.PortForwarder
+	fw, err = portforward.NewOnAddresses(dialer, []string{"localhost"}, portPairs, stopChan, nil, nil, errOut)
+	if err != nil {
+		return err
+	}
+
+	// start port-forwarding
+	go func() {
+		err = fw.ForwardPorts()
+		if err != nil {
+			fmt.Fprint(out, fmt.Errorf("error setting up port forwarding: %v", err).Error())
+			os.Exit(1)
+		}
+	}()
+
+	printPortForwardingInfo(portPairs, out)
+	return nil
+}
+
+func printPortForwardingInfo(portPairs []string, out io.Writer) {
```

Comment:
Please prepare for `json` output, this function should return a structured data instead of writing to a buffer

Reply:
What should it return to? Once inside `pkg/dev/dev.go`, we have long-running tasks which don't return to CLI layer till the user hits Ctrl+C.

Reply:
You could pass a function instead of the writer, and call this function with the structured data

Reply:
> You could pass a function instead of the writer, and call this function with the structured data

@feloy TBH I'm still struggling to understand what you are asking for here. 😕

Reply:
@dharmit It is about using a handler, like we discussed earlier today